% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AnnotationHelper.R
\name{select.annotations}
\alias{select.annotations}
\title{Annotations selection helper.}
\usage{
select.annotations(genome.build)
}
\arguments{
\item{genome.build}{The genome build to use for annotation.}
}
\value{
A list with the following elements: \describe{
  \item{TxDb} {A TxDb of class \linkS4class{AnnotationDBI} providing informations about the genes
  of the selected build.}
  \item{OrgDb} {An \linkS4class{OrgDb} object for the selected species.}
  \item{OrgDbStr} {A character string representing the name of the \linkS4class{OrgDb} object.}
  \item{BSGenome} {A \linkS4class{BSGenome} object to retrieve DNA sequences.}
  \item{KEGG} {A cache of KEGG pathways, as returned by \code{\link[gage]{kegg.gsets}} from the \pkg{gage} library.}
  \item{PWMBG} {A PWMLogn background for motif enrichment of promoter regions.}}
}
\description{
Given a genome build identifier, creates a list of annotation databases
which can be passed to the annotation functions.
}

